base/array.jl

```diff
@@ -1798,6 +1801,9 @@ findfirst(p::Union{Fix2{typeof(isequal),T},Fix2{typeof(==),T}}, r::AbstractUnitR
     first(r) <= p.x <= last(r) ? 1+Int(p.x - first(r)) : nothing
 
 function findfirst(p::Union{Fix2{typeof(isequal),T},Fix2{typeof(==),T}}, r::StepRange{T,S}) where {T,S}
+    if RangeStepStyle(r) !== RangeStepRegular()
```

Comment:
Would it be worth making this a trait-depot instead?
```julia
findfirst(p::Union{Fix2{typeof(isequal),T},Fix2{typeof(==),T}}, r::StepRange{T,S}) where {T,S} = _findfirst(p, RangeStepStyle(r), r)
_findfirst(p, ::RangeStepStyle, r) = invoke(findfirst, Tuple{Function,Any}, p, r)  # fallback
function _findfirst(p, ::RangeStepRegular, r)
    ...
end
```
That would make it possible to specialize at intermediate states.

Reply:
That looks neater. Should I perhaps also change the simplest findfirst method to be `findfirst_iter` or something, so that it can be called without `invoke`?

Reply:
See if you like what I did here. I made:
```julia
_findfirst(p, ::RangeStepStyle, r) = _findfirst(p, r)
```
i.e. putting the trait before the variable it refers to, and breaking up the definition of the simplest `findfirst(p, r)` so that I can call its implementation without invoke.